### PR TITLE
feat: extension system

### PR DESCRIPTION
**Pull request by zachfedor**
Architecture for #160 

First off, the Stellar CLI plugin system isn't really a plugin system. It's a naming convention to allow discovery of 3rd party commands (like Scaffold) and allow $PATH-based command calling via the parent `stellar` command. There's no shared context, no lifecycle hooks, and no communication between commands. Scaffold extensions, on the other hand, probably need to be able to receive pipeline state (contract addresses, WASM paths, network config, etc) **AND** react to changes to them via events during our `watch` loop.

So we can still use the Stellar plugin naming convention (`stellar-scaffold-<ext-name>`) that way we can use `stellar plugin` sub-commands, but we'll need a deeper integration. Installation is via `cargo install stellar-scaffold-<ext-name>` and Scaffold discovers them on PATH at startup based on the project's configuration (i.e. an ordered array of extensions).

# Proposed Lifecycle Hooks
- ~`init`: `pre-init` → `post-init`~
- `build`: `pre-build` → `post-build` → `pre-deploy` → `post-deploy` → `pre-codegen` → `post-codegen`
- `watch`: `pre-watch` → `post-watch`

# Design
So the question is how do we run an extension at hook time? The simplest option is as sub-commands for each lifecycle hook. Context can be passed to each sub-command as arguments or even as JSON piped on stdin (e.g. `stellar-scaffold-my-ext post-deploy < context.json`).

Extensions could even provide a `manifest` command that returns which hooks it supports so we only call an extension as needed and they only need to implement the sub-commands they care about.

## Long-Lived Extensions
For something like an indexer plugin, it would need its own background process during our `watch` command. It can setup and teardown that process on the pre/post watch hooks and then manage its own process as needed on any subsequent hooks.

# Configuration
I know we're changing the `environments.toml` file eventually (#181), but for lack of a better place to put it, we can configure the order of extensions to run in there. We can also allow optional per-extension configuration there too, which might change per environment. For example:

```toml
[development]
extensions = ["reporter", "indexer"]

[development.extensions.reporter]
warn_size_kb = 128

[development.extensions.indexer]
storage = "sqlite"
events = ["transfer", "mint"]
api_port = 3100
```
Scaffold resolves each name to stellar-scaffold-<name> on PATH, calls manifest to validate, then invokes lifecycle sub-commands in array order at each pipeline stage. 

# Open Questions
1. Should Scaffold wait for the subprocess to exit before proceeding, or fire-and-forget for some hooks? Configurable per-hook in the manifest?
2. Should `pre-watch` be able to signal that it spawned a background process, so Scaffold knows to call `post-watch` on exit?

# Tasks
- [X] Create a crate to define types needed for extensions (e.g. manifest output, hook context values, etc)
- [X] Add extension config parsing to environments.toml
- [X] Add an extension discovery mechanism to validate extensions registered in config
- [X] Create a hook runner to call each extensions subcommand and pass appropriate context
- [X] Wire up the hook runner within `build` pipeline
- [X] Wire up the hook runner within `watch` pipeline
- [X] Add `stellar scaffold ext ls` command to read extension config, resolve their path, call their manifest, and print version, status, and supported hooks
- [ ] Create the `stellar-scaffold-reporter` example extension as reference extension and for end-to-end tests

### CODE REVIEW

File: crates/stellar-scaffold-cli/src/commands/build/clients.rs [hunk 1]
```diff
@@ -273,17 +309,41 @@ export default new Client.Client({{
         Ok(())
     }
 
-    async fn generate_contract_bindings(&self, name: &str, contract_id: &str) -> Result<(), Error> {
+    #[allow(clippy::too_many_lines)]
```
Comment (zachfedor) [inline crates/stellar-scaffold-cli/src/commands/build/clients.rs]: Same as the above comment, but this time for `clippy:too_many_lines`. Probably a good way to refactor, but I didn't want to potentially change behavior while adding extensions.

File: crates/stellar-scaffold-cli/src/commands/build/mod.rs [hunk 2]
```diff
@@ -115,16 +117,81 @@ impl Command {
...
+            &extensions,
+            HookName::PreCompile,
+            &pre_compile_ctx,
+            &printer,
+        )
+        .await;
+
         for p in &packages {
             self.create_cmd(p, target_dir)?.run(global_args)?;
         }
```
Comment (chadoh) [inline crates/stellar-scaffold-cli/src/commands/build/mod.rs]: Am I understanding correctly that pre- and post-compile hooks run before and after _all_ packages are compiled, but pre- and post-codgen are run surrounding codegen for _each_ package?

File: crates/stellar-scaffold-cli/src/commands/build/env_toml.rs [hunk 3]
```diff
@@ -170,3 +234,129 @@ impl From<&Network> for stellar_cli::config::network::Args {
...
+[development.ext.unlisted]
+some_key = "value"
+"#
+        );
+
+        let env = parse_dev(&toml);
+        assert_eq!(env.extensions.len(), 1);
+        assert_eq!(env.extensions[0].name, "reporter");
+    }
+}
```
[the remark this answers is not in the document]
Comment (zachfedor) [reply crates/stellar-scaffold-cli/src/commands/build/env_toml.rs]: That's a feature! You can remove an extension from the array to prevent it from running but leave the configuration in place if you want to turn it back on later.

[the remark this answers is not in the document]
Comment (chadoh) [reply crates/stellar-scaffold-cli/src/commands/build/env_toml.rs]: Yes, my point is that the current test checks:

- if `[development.extensions]` length == 1
- and two separate `[development.ext.whatever]` sections
- no bugs, only the one extension included

We don't currently have a test for:

- if `[development.extensions]` is empty
- but ≥1 `[development.ext.whatever]` tables

nor do we have a test for:

- no `[development.extensions]` setting at all
- but ≥1 `[development.ext.whatever]` tables

Maybe these are not worth testing? But I could imagine authoring things in such a way that the current test passes but one of these other situations would be a bug-causing edge case
